src/actions/2fa/attach.js

```diff
@@ -0,0 +1,55 @@
+const { ActionTransport } = require('@microfleet/core');
+const Promise = require('bluebird');
+const redisKey = require('../../utils/key');
+const handlePipeline = require('../../utils/pipelineError');
+const hasTotp = require('../../utils/hasTotp.js');
+const { verifyTotp, generateRecoveryCodes } = require('../../utils/2fa.js');
+const { USERS_2FA_SECRET, USERS_2FA_RECOVERY } = require('../../constants');
+
+function storeData(userId, secret, recoveryCodes) {
+  // store secret key and recovery codes
+  return this.redis
+    .pipeline()
+    .set(redisKey(USERS_2FA_SECRET, userId), secret)
+    .sadd(redisKey(USERS_2FA_RECOVERY, userId), recoveryCodes)
+    .exec()
+    .then(handlePipeline)
+    .return({ recoveryCodes, enabled: true });
+}
+
+/**
+ * @api {amqp} <prefix>.attach Attach
+ * @apiVersion 1.0.0
+ * @apiName Attach
+ * @apiGroup Users
+ *
+ * @apiDescription Allows to attach secret key and recovery code to user's account,
+ * generate and returns initial recovery code.
+ *
+ * @apiHeader (Authorization) {String} Authorization JWT :accessToken
+ * @apiHeaderExample Authorization-Example:
+ *     "Authorization: JWT my.reallyniceandvalid.jsonwebtoken"
+ * @apiHeader (Authorization) {String} X-Auth-TOTP TOTP or recoveryCode
+ * @apiHeaderExample X-Auth-TOTP-Example:
+ *     "X-Auth-TOTP: 123456"
+ *
+ * @apiParam (Payload) {String} username - id of the user
+ * @apiParam (Payload) {String} secret - crypto secure 32 characters hex key
```

Comment:
I don't think that a `secret` should be accepted as an input parameter, instead it should be generated internally via [2fa.generateKey](https://github.com/simontabor/2fa/blob/master/lib/2FA.js#L8) or [optlib-utils/secretKey](https://yeojz.github.io/otplib/docs/module-otplib-utils_secretKey.html)

Reply:
Well, it must be, because the flow is the following:

1. ask for "generateKey" -> show QR to the user
2. get the "totp" and associated secret -> attach it

One possible solution is to make "attaching" stateful, but I think its just a waste of time, because a) user must be authenticated to attach 2FA b) user doesn't want to "fake" a secret code, because it reduces the specific user's security only